Feature/implement profilepic enhancements

raniabbrk opened the pull request:
**Description**
This pull request introduces significant updates to the profile and avatar selection features, improving UI/UX, adding new functionality, and enhancing test coverage. It includes refactoring for better maintainability, localization support, and rigorous testing to ensure high-quality user interaction.

---

 **Changes Introduced**

### 1. **Profile Screen Updates**
   - Refactored `ProfileScreen` for better alignment with Material3 design principles.
   - Enhanced UI with dynamic buttons: There is now a Floating Action Button ("+") that appears next to the avatar when it is set to default (i.e, the user does not have an avatar). When the user selects an avatar other than default one, then the "Change Avatar" button appears  instead of the Floating Action Button.
---
### 2. **Avatar Selection Screen Enhancements**

- Improved the `AvatarSelectionScreen` for a smoother and more intuitive user experience:
  - **"Reset to Default Avatar" Button**: Added a button to allow users to reset to the default avatar (no avatar). This button is now:
    - Disabled when the user is already using the default avatar.
  - **"Confirm Selection" Button**: Disabled when the user has not selected a new avatar.

- **Navigation Update**: Added a "Back Arrow" to let users return to the profile screen when neither button is clickable, allowing them to exit without performing any action.

- **Visual Update**: Avatars are now displayed within circular frames, providing a clean and visually distinct design.

- **Design Upgrade**: Integrated Material3 design principles, including updated button styles and hover effects.

- **Preview without Immediate Commitment**
  - **`pendingAvatar` **: Unlike the old implementation where selecting an avatar immediately updated the user profile, `pendingAvatar` decouples selection from confirmation. 
  - **Flexibility**: Users can change their selection multiple times without triggering profile updates until they explicitly confirm.
  - **Reduced API Calls**: Since the profile is only updated upon confirmation, unnecessary API calls for intermediate selections are avoided, improving efficiency.
 ---

### 3. **AvatarViewModel Improvements**
   - Refactored ViewModel to support nullable avatars.
  
---
### 4. **Localization Support**
   - Added new resource strings for all user-facing texts in `res/values/strings.xml`.
   - Externalized button texts, content descriptions, and navigation labels for easier translation.
   - Ensured that all UI components use `stringResource` for localization compatibility.
---
### 5. **Test Coverage Expansion**
   - Comprehensive updates to `AvatarSelectionScreenTest`:
     - Added scenarios for avatar pre-selection, reset and confirm buttons behavior, and navigation.
   - Updated `ProfileKtTest`:
     - Verified dynamic button behaviors (e.g., FAB visibility).



---

 **Related Issues**
#221

raniabbrk committed:
refactor: enhance avatar selection screen

- Replaced older components with Material3 design elements.
- Introduced `pendingAvatar` state for better selection management.
- Added shadow and background for better visual appeal.
- Enhanced button functionality with state-dependent styling and actions.

raniabbrk committed:
refactor: improve AvatarViewModel functionality

- Modified `saveSelectedAvatar` and `updateUserProfileWithAvatar` to handle nullable avatar IDs.

raniabbrk committed:
refactor: enhance ProfileScreen layout and functionality

- Added `FloatingActionButton` for quick avatar selection when no avatar is chosen.
- Incorporated Material3 design elements for buttons and icons.
- Improved alignment and spacing for better layout adaptability in different orientations.
- Utilized `stringResource` for content descriptions and button texts to support localization.
- Adjusted visual hierarchy with conditionally rendered "Change Avatar" button.

raniabbrk committed:
refactor: improve ProfileRepository functionality
- Modified `saveSelectedAvatar` and `updateUserProfileWithAvatar` to handle nullable avatar IDs.

raniabbrk committed:
refactor: improve ProfileRepositoryFirestore functionality
- Modified `saveSelectedAvatar` and `updateUserProfileWithAvatar` to handle nullable avatar IDs.

raniabbrk committed:
test: enhance AvatarSelectionScreenTest with additional scenarios

- Updated `AvatarSelectionScreenTest` with expanded test cases.
- Added tests for:
  - Pre-selecting an existing avatar.
  - Reset button behavior for default and custom avatars.
  - Floating Action Button (FAB) visibility for avatar selection.
  - Confirm button functionality for avatar update and navigation.
  - Go-back arrow display and navigation.
- Integrated `getResourceString` for text-based assertions to support localization.
- Improved mock interactions for better validation of repository calls.

raniabbrk committed:
test: enhance AvatarViewModelTest with additional scenarios
- Modified `saveSelectedAvatar` to handle nullable avatar IDs.

raniabbrk committed:
test: improve ProfileKtTest for enhanced validation and functionality

- Added `getResourceString` usage for localization support in assertions.
- Improved test coverage with new scenarios:
  - Floating Action Button (FAB) visibility and navigation for avatar selection.
  - Validation of landscape orientation and scrollability.
- Refined mocks for `NavigationActions` and `AvatarViewModel` to ensure accurate state handling.
- Enhanced assertions for buttons and bottom navigation tabs to verify proper rendering and interactions.

raniabbrk committed:
feat: add resource strings for improved localization

- Added new strings in the `res/values/strings.xml`

raniabbrk committed:
refactor: remove deprecated `verifyOrCreateProfile` reference

- Eliminated outdated `verifyOrCreateProfile` usage in `AvatarSelectionScreen.kt`.
- Aligned code with current implementation

raniabbrk committed:
chore: Resolve merge conflicts

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/ui/profile/Profile.kt:
```diff
@@ -63,44 +66,64 @@ fun ProfileScreen(navigationActions: NavigationActions, avatarViewModel: AvatarV
-                    }
+                  // Show FAB when no avatar is selected
+                  if (selectedAvatar == null || selectedAvatar == R.drawable.default_profile_icon) {
+                    FloatingActionButton(
+                        onClick = { navigationActions.navigateTo(Screen.AVATAR_SELECTION) },
+                        containerColor = DarkPurple,
+                        modifier =
+                            Modifier.size(36.dp) // Smaller size for Instagram-like style
+                                .offset(x = (-8).dp, y = (-8).dp) // Adjust position slightly
+                        ) {
+                          Icon(
+                              imageVector = Icons.Filled.Add,
+                              contentDescription = stringResource(R.string.add_avatar),
+                              tint = Color.White)
+                        }
```
You could consider wrapping this code in a small composable function to make it clear that it represents the Floating Action Button (FAB) for adding an avatar. This would improve readability and make the purpose of the code more evident.

raniabbrk replied:
Great idea, thank you for this suggestion.

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/ui/profile/Profile.kt:
```diff
@@ -63,44 +66,64 @@ fun ProfileScreen(navigationActions: NavigationActions, avatarViewModel: AvatarV
-                      Text("Change Avatar")
-                    }
+                  // Show FAB when no avatar is selected
+                  if (selectedAvatar == null || selectedAvatar == R.drawable.default_profile_icon) {
+                    FloatingActionButton(
+                        onClick = { navigationActions.navigateTo(Screen.AVATAR_SELECTION) },
+                        containerColor = DarkPurple,
+                        modifier =
+                            Modifier.size(36.dp) // Smaller size for Instagram-like style
+                                .offset(x = (-8).dp, y = (-8).dp) // Adjust position slightly
+                        ) {
+                          Icon(
+                              imageVector = Icons.Filled.Add,
+                              contentDescription = stringResource(R.string.add_avatar),
+                              tint = Color.White)
+                        }
+                  }
+                }
+
+                if (selectedAvatar != null && selectedAvatar != R.drawable.default_profile_icon) {
+                  // Show "Change Avatar" button only if a custom avatar is selected
+                  Button(
+                      onClick = { navigationActions.navigateTo(Screen.AVATAR_SELECTION) },
+                      modifier = Modifier.padding(top = 16.dp),
+                      colors = ButtonDefaults.buttonColors(DarkPurple)) {
+                        Text(stringResource(R.string.change_avatar))
+                      }
+                }
```
Wouldn't it be better to use an `if-else` condition here instead of two `if` statements? This would make the logic clearer and help avoid potential issues with overlapping conditions.

raniabbrk replied:
Thank you for pointing out the lack of clarity. But due to the layout structure, specifically with the Box and Column, implementing an if-else statement for the conditional rendering of the FAB and the "Change Avatar" button would not work as intended. This is because the Box and Column require independent control over their child elements, and merging the logic into an if-else would result in layout inconsistencies.
Instead, I opted to refactor the conditional logic by extracting the FAB and button into their respective composable functions, ProfileFab and ChangeAvatarButton. I think this approach keeps the logic modular and enhances code readability.

AdrienBousquieEPFL reviewed (dismissed):
The feature is really cool! I tested it on the phone and didn’t encounter any issues. 

I don’t have the time to review everything thoroughly right now but will come back to it later. In the meantime, please address the mentioned points. Additionally, the Figma design hasn’t been updated yet—please ensure it’s updated before merging. Thank you!

raniabbrk committed:
refactor: improve conditional rendering for Profile FAB and Change Avatar button

- Extracted FAB rendering into a reusable ProfileFab composable
- Extracted Change Avatar button logic into ChangeAvatarButton composable
- Enhanced readability and modularity of ProfileScreen

raniabbrk committed:
refactor: extract FloatingActionButton logic into AddAvatarFab composable

- Encapsulated FloatingActionButton into a reusable AddAvatarFab composable
- Improved modularity and reusability of FAB components across screens

raniabbrk committed:
refactor: add prefixes to string resources for better organization

- Updated string resource names to include the screen prefix
- Refactored references in the screens to use the new string names
- Improves maintainability and consistency in string naming

raniabbrk committed:
refactor: add prefixes to string resources for better organization

- Updated string resource names to include the screen prefix
- Refactored references in the screens to use the new string names
- Improves maintainability and consistency in string naming

raniabbrk committed:
feat: add reusable BackgroundImage composable for screen backgrounds

- Created a reusable BackgroundImage composable with painter resource and content description
- Added support for test tags to facilitate UI testing

raniabbrk commented on app/src/main/java/com/github/lookupgroup27/lookup/ui/profile/Profile.kt:
```diff
@@ -53,7 +56,7 @@ fun ProfileScreen(navigationActions: NavigationActions, avatarViewModel: AvatarV
-              contentDescription = "Background",
+              contentDescription = stringResource(R.string.background_description),
               contentScale = ContentScale.Crop,
               modifier = Modifier.fillMaxSize())
```
You're right, thank you for noticing that.

raniabbrk committed:
fix: correct background test tag string to resolve failing quiz tests

- Updated the test tag string from "Background" to "background_test_tag" in the strings file
- Ensured consistency with expected test tag usage in the quiz tests
- Fixes failing UI tests related to background identification
g

raniabbrk committed:
fix: update test tags to align with modified string resources and resolve test failures

rihabbelmekki reviewed (approved):
Great work on this PR! This feature significantly enhances the user experience, and you've effectively implemented the requested changes and added improvements to the profile and avatar selection features. Everything looks solid to me—approved!

Ismaillat reviewed (changes requested):
The code implements good changes and effectively ensures clear functionality and intuitive UI behavior. There are small changes that could improve code.
Good Job !

raniabbrk commented on app/src/main/java/com/github/lookupgroup27/lookup/ui/profile/components/AddAvatarFab.kt:
```diff
@@ -0,0 +1,34 @@
+ *   the avatar selection screen or to trigger avatar addition logic.
+ */
+@Composable
+fun AvatarFab(onClick: () -> Unit) {
```
You're write, thank you for noticing that!

raniabbrk committed:
refactor: Improve avatar handling and logging across components
- Added edge case handling for invalid or default selectedAvatar values
- Introduced isAvatarInvalid and isAvatarDefaultOrNull variables for better readability
- Added debug logging for invalid selectedAvatar in ChangeAvatarButton and other components
- Refactored conditional logic to simplify and improve clarity

Ismaillat reviewed (approved):
The code looks good to me and is ready to be merged.
